### PR TITLE
Update usage examples in README

### PR DESCRIPTION
**Pull request by mico**
Related #166 
I found a bug related to updated documentation (#167), it will be fixed separately.

### CODE REVIEW

**Review by ronaldtse — CHANGES_REQUESTED** [review]
Instead of changing the README, can you make this also work? `Pubid::Iso::Identifier.new(publisher: "ISO", number: 123)`

**Comment by mico** [comment]
> Instead of changing the README, can you make this also work? `Pubid::Iso::Identifier.new(publisher: "ISO", number: 123)`

It's still working (using `Pubid::Iso::Identifier::Base.new`), but not recommended because you can't apply identifier type there.
We are using `#create` to resolve identifier type. Identifier type could be provided through "type" (:tr) parameter or could be part of "typed_stage" ("DTR").

**Comment by ronaldtse** [comment]
@mico there is a real problem here where the API expectations across different `Pubid` flavors differ.

1. People are already using `Pubid::Iso::Identifier.new`. We cannot "remove the method" and ask people to change.
2. People are also using `Pubid::Iec::Identifier.new` and `Pubid::Nist::Identifier.new`. We have to maintain the same API pattern across flavors. We cannot ask users to only use `Pubid::Iso::Identifier::Base.new` for ISO and then `Pubid::Iec::Identifier.new` for IEC.

The Pubid::Iso API has already been massively changed multiple times since it has been used: @opoudjis and @andrew2net have trouble following those changes, and it really shouldn't happen.

Can you make `Pubid::Iso::Identifier.new` work? If not, please suggest a reasonable replacement that will work across ALL flavors consistently and will never change again. Thanks.

**Comment by mico** [comment]
> @mico there is a real problem here where the API expectations across different `Pubid` flavors differ.
> 
> 1. People are already using `Pubid::Iso::Identifier.new`. We cannot "remove the method" and ask people to change.
> 2. People are also using `Pubid::Iec::Identifier.new` and `Pubid::Nist::Identifier.new`. We have to maintain the same API pattern across flavors. We cannot ask users to only use `Pubid::Iso::Identifier::Base.new` for ISO and then `Pubid::Iec::Identifier.new` for IEC.
> 
> The Pubid::Iso API has already been massively changed multiple times since it has been used: @opoudjis and @andrew2net have trouble following those changes, and it really shouldn't happen.
> 
> Can you make `Pubid::Iso::Identifier.new` work? If not, please suggest a reasonable replacement that will work across ALL flavors consistently and will never change again. Thanks.

If we use `Pubid::Iso::Identifier.new` then identifier's object will always be `Pubid::Iso::Identifier` class. It's impossible to return another class (`InternationalStandard` or `TechnicalReport`) without factory method (#create).
But I can work on storing real identifier's object inside `Pubid::Iso::Identifier` and delegate their values to `Pubid::Iso::Identifier`. Do you think it could work as a solution?

**Comment by ronaldtse** [comment]
Ping @opoudjis for input here.

**Comment by opoudjis** [comment]
I'm a bit confused as to why Ronald has asked me to comment, since I'm only going to agree with him.

> If we use Pubid::Iso::Identifier.new then identifier's object will always be Pubid::Iso::Identifier class. It's impossible to return another class (InternationalStandard or TechnicalReport) without factory method (#create).

... Why the HELL are InternationalStandard and TechnicalReport not subclasses of Pubid::Iso::Identifier, to begin with?!

> Pubid::Iso::Identifier::Base.parse("ISO/FDIS 26000:2010(ru)")

I don't understand why I have to add "Base", and when I don't add "Base". Nothing in your readme explains why I have to add "Base", and when I don't add "Base".

And I don't want an explanation, really, because I should NOT have to be bothered with this. If you bother me with your internal state, then you are failing to abstract  your internal state away from me. For example, I assume, the fact that you have modelled  Iso::Identifier::TechnicalReport as a subclass of Iso::Identifier:Base instead of as a subclass of Iso::Identifier. I should not have to care how you've modelled them. Just as I should never have had to be exposed to Iso::Identifier::Status in order to represent document status.

If you are not abstracting your internal state away from me, that means you are not doing your job. The ENTIRE POINT of object oriented programming is that you hide details from outside users, and make things easier for them.

I want to call ONE method, applicable throughout ISO, that takes a bunch of stable parameters, and gives me a correctly formatted identifier.

If you can guarantee that a call to Pubid::Iso::Identifier::Base will be used in **absolutely all instances** where an ISO identifier is called for, and that the same will apply for IEEE or NIST or IEC, then I will put up with it as an annoying eccentricity.

But if it turns up that I **ever** need to switch Base to TechnicalReport or InternationalStandard, depending on the type of identifier, then you are making my life unnecessarily complex, and I reject your solution.

And right now,  I do not know which of the two is the case. Your documentation suggests the former, e.g. `pubid = Pubid::Iso::Identifier::Base.parse("ISO/WD TR 26000:2010")`. But you need to be explicit about (1) what subclasses exist (because developers are going to run into them in usage), and (2) whether developers ever need to use them separately from Base, or whether all their methods are the same.

-----

If you change the interface every few months, you do not make things easier for me. You make me refuse to use your gems until the interface has stabilised. (Which is what has happened. Andrej is in fact using your gem; I have a PR that I will not update, let alone merge, until I know that you are not changing the call signature any more, and you still are changing it.) 

And I blame Ronald as well as you for the instability of the interface: Ronald wanted you to fix the interface to make it more logical, but in a SDK or API invoked by an external consumer, stability and consistency are FAR MORE IMPORTANT than logicalness.

This gem has been under development for at least 6 months, and to be blunt, it is not a 6-month job. I do NOT care how you implement it. I do care that anything any of us say now is going to delay this gem finishing even longer.

If you've gotten yourself in such a mess with your internal modelling that you cannot expose Pubid::Iso::Identifier.new any more, then fine, expose  Pubid::Iso::Identifier.Base.new instead, so long as that is the ONLY way to create ISO identifiers. If you cannot just use Pubid::Iso::Identifier.Base.new either (and I have no idea whether that's true or not), then I dunno, make up a new method like  Pubid::Iso::Identifier.create, to hide the detail from me. I don't care any more. (Though Andrej will, because unlike me he is already using the class in production.) 

But whatever you do, if you change it now, you MUST ONLY CHANGE IT ONCE. I have no interest in updating my code every few months for something that is meant to be removing work from me.

**Comment by ronaldtse** [comment]
Following up on @opoudjis 's rant, I'd like to clarify that there are TWO types of usages.

## Usage 1: Parsing an existing identifier

```ruby
Pubid::Iso::Identifier.parse("ISO/WD TR 26000:2010")
# => Pubid::Iso::Identifier::TechnicalReport
```

This is what @opoudjis has been indicating at. And this can be implemented fine, there is no problem on @mico 's side here.

## Usage 2: Building an identifier

The problem is this:
```ruby
Pubid::Iso::Identifier.new("ISO/WD TR 26000:2010")
# => Pubid::Iso::Identifier ...
```

As @mico alluded to, the `#new` method in Ruby requires returning an instance of the same class of the method that is called, which is `Pubid::Iso::Identifier`.

Now, whether `Pubid::Iso::Identifier::TechnicalReport` should be a subclass of `Pubid::Iso::Identifier` can be subject of debate -- but in Ruby, inheriting a parent module causes namespace issues (modules mixed into the parent module will be unintentionally inherited by the subclasses). The usage of `Base` is a common pattern: for example, ActiveRecord in Rails chose to use `ActiveRecord::Base` instead.

In my experience, the usage of `Base` avoids mysterious namespace issues, and is the way to go.

Therefore, in this pattern, this cannot be done:

```ruby
Pubid::Iso::Identifier.new("ISO/WD TR 26000:2010")
# => Pubid::Iso::Identifier::TechnicalReport
Pubid::Iso::Identifier.new(publisher: "ISO", stage: "WD", type: "TR", number: 26000, year: "2010")
# => Pubid::Iso::Identifier::TechnicalReport
```

This however works well:

```ruby
Pubid::Iso::Identifier::TechnicalReport.new("ISO TR 26000:2010")
# => Pubid::Iso::Identifier::TechnicalReport
Pubid::Iso::Identifier::TechnicalReport.new(publisher: "ISO", stage: "WD", number: 26000, year: "2010")
# => Pubid::Iso::Identifier::TechnicalReport
```

Of course, it also means @mico can add an auxiliary function (such as `new_with_type`) to help with this:
```ruby
Pubid::Iso::Identifier.new_with_type(type: :tr, publisher: "ISO", stage: "WD", number: 26000, year: "2010")
# => Pubid::Iso::Identifier::TechnicalReport
```

This means that when Metanorma builds identifiers, it will need to do this:

```ruby
Pubid::Iso::Identifier::TechnicalReport.new(publisher: "ISO", stage: "WD", number: 26000, year: "2010")
# => Pubid::Iso::Identifier::TechnicalReport
```

Or:
```ruby
Pubid::Iso::Identifier.new_with_type(type: :tr, publisher: "ISO", stage: "WD", number: 26000, year: "2010")
# => Pubid::Iso::Identifier::TechnicalReport
```

## Conclusion

@opoudjis are you okay with the approaches described above (1 and 2)?

**Comment by mico** [comment]
> Of course, it also means @mico can add an auxiliary function (such as `new_with_type`) to help with this:
> 
> ```
> Pubid::Iso::Identifier.new_with_type(type: :tr, publisher: "ISO", stage: "WD", number: 26000, year: "2010")
> # => Pubid::Iso::Identifier::TechnicalReport
> ```

We already have "an auxiliary function" `#create`.
```
Pubid::Iso::Identifier::Base.create(type: :tr, publisher: "ISO", stage: "WD", number: 26000, year: "2010")
# => Pubid::Iso::Identifier::TechnicalReport
```

But I can move `Pubid::Iso::Identifier::Base.create` to `Pubid::Iso::Identifier.create`

**Comment by opoudjis** [comment]
Fine

**Comment by ronaldtse** [comment]
Thanks @opoudjis -- @mico can you help make the changes to allow both Usages 1 and 2? Thanks.

**Comment by mico** [comment]
> Thanks @opoudjis -- @mico can you help make the changes to allow both Usages 1 and 2? Thanks.

@ronaldtse done

File: README.adoc [hunk 1]
```diff
@@ -18,7 +18,7 @@ identifiers.
 ----
 require "pubid-iso"
 
-pubid = Pubid::Iso::Identifier.new(publisher: "ISO", number: 123)
+pubid = Pubid::Iso::Identifier::Base.create(publisher: "ISO", number: 123)
```
Comment (ronaldtse) [inline README.adoc:21]: Update the README to the new method location?

**Review by ronaldtse — CHANGES_REQUESTED** [review]

File: lib/pubid/iso/identifier.rb [hunk 2]
```diff
@@ -0,0 +1,55 @@
+module Pubid::Iso
+  module Identifier
+
+    # Resolve identifier's class and create new identifier
+    # @see Pubid::Identifier::Base.initialize for available options
+    def Identifier.create(**opts)
```
[the remark this answers is not in the document]
Comment (mico) [reply lib/pubid/iso/identifier.rb]: Fixed

**Review by ronaldtse — APPROVED** [review]
Thank you!

**Pull request merged**